server/src/main/java/org/opensearch/search/startree/StarTreeFilter.java

```diff
@@ -0,0 +1,277 @@
+
+        // Use BFS to traverse the star tree
+        Queue<StarTreeNode> queue = new ArrayDeque<>();
+        queue.add(starTree);
+        int currentDimensionId = -1;
+        Set<String> remainingPredicateColumns = new HashSet<>(_predicateEvaluators.keySet());
+        if (foundLeafNode) {
+            globalRemainingPredicateColumns = new HashSet<>(remainingPredicateColumns);
+        }
+
+        int matchedDocsCountInStarTree = 0;
+        int maxDocNum = -1;
+
+        StarTreeNode starTreeNode;
+        List<Integer> docIds = new ArrayList<>();
+        while ((starTreeNode = queue.poll()) != null) {
+            int dimensionId = starTreeNode.getDimensionId();
+            if (dimensionId > currentDimensionId) {
+                // Previous level finished
+                String dimension = dimensionNames.get(dimensionId);
+                remainingPredicateColumns.remove(dimension);
+                if (foundLeafNode && globalRemainingPredicateColumns == null) {
+                    globalRemainingPredicateColumns = new HashSet<>(remainingPredicateColumns);
+                }
+                currentDimensionId = dimensionId;
+            }
+
+            // If all predicate columns are matched, we can use aggregated document
+            if (remainingPredicateColumns.isEmpty()) {
+                int docId = starTreeNode.getAggregatedDocId();
+                docIds.add(docId);
+                matchedDocsCountInStarTree++;
+                maxDocNum = Math.max(docId, maxDocNum);
+                continue;
+            }
+
+            // For leaf node, add the range of documents for this node to the list
+            if (starTreeNode.isLeaf()) {
+                for (long i = starTreeNode.getStartDocId(); i < starTreeNode.getEndDocId(); i++) {
+                    docIds.add((int) i);
+                    matchedDocsCountInStarTree++;
+                    maxDocNum = Math.max((int) i, maxDocNum);
+                }
+                continue;
+            }
+
+            // For non-leaf node, proceed to next level
+            String childDimension = dimensionNames.get(dimensionId + 1);
+
+            // Only read star-node when the dimension is not in the global remaining predicate columns
+            StarTreeNode starNode = null;
+            if (globalRemainingPredicateColumns == null || !globalRemainingPredicateColumns.contains(childDimension)) {
+                starNode = starTreeNode.getChildForDimensionValue(StarTreeUtils.ALL, true);
+            }
+
+            if (remainingPredicateColumns.contains(childDimension)) {
+                // Calculate the matching dictionary ids for the child dimension
+                int numChildren = starTreeNode.getNumChildren();
+
+                // Get the query ranges for the child dimension
+                List<Range> queryRanges = _predicateEvaluators.get(childDimension);
+
+                Iterator<? extends StarTreeNode> childrenIterator = starTreeNode.getChildrenIterator();
```

Comment:
Lets avoid using getChildrenIterator since it iterates through all nodes, we should do binary search.

@sarthakaggarwal97 and i were discussing the algorithm and better way is to do.


We can create an iterator in FixedLengthStarTreeNode based on the following logic :

```
public Iterator<FixedLengthStarTreeNode> getChildrenIteratorForRange(int min, int max) throws IOException {
        return new Iterator<>() {
            private int currentChildId = binarySearchChildNodeId(min);
            private final int lastChildId = binarySearchChildNodeId(max);

            @Override
            public boolean hasNext() {
                return currentChildId <= lastChildId;
            }

            @Override
            public FixedLengthStarTreeNode next() {
                try {
                    return new FixedLengthStarTreeNode(in, currentChildId++);
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            }

            @Override
            public void remove() {
                throw new UnsupportedOperationException();
            }
        };
    }
```

Reply:
Thanks for suggestions. Made those changes.